pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_lifecycle.go

```diff
@@ -0,0 +1,115 @@
+func (r *VirtualNodeReconciler) deleteNamespaceMap(nm mapsv1alpha1.NamespaceMap) error {
+
+	nm.SetFinalizers(slice.RemoveString(nm.GetFinalizers(), namespaceMapFinalizer, nil))
+	klog.Info("Someone try to delete namespaceMap, ok delete!!")
+
+	if err := r.Update(context.TODO(), &nm); err != nil {
+		klog.Error(err, " --> Problems while removing NamespaceMap's finalizer")
+		return err
+	}
+	return nil
+}
+
+// create a new NamespaceMap with Finalizer and OwnerReference
+func (r *VirtualNodeReconciler) createNamespaceMap(n corev1.Node) error {
+	klog.Infof(" Create NamespaceMap for %s", n.Name)
+	nm := &mapsv1alpha1.NamespaceMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "namespaceresources.liqo.io/v1",
+			Kind:       "NamespaceMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", n.GetAnnotations()[const_ctrl.VirtualNodeClusterId]),
+			Namespace:    const_ctrl.MapNamespaceName,
+			Labels: map[string]string{
+				const_ctrl.VirtualNodeClusterId: n.GetAnnotations()[const_ctrl.VirtualNodeClusterId],
+			},
+		},
+	}
+
+	nm.SetFinalizers(append(nm.GetFinalizers(), namespaceMapFinalizer))
+
+	nm.SetOwnerReferences(append(nm.OwnerReferences, metav1.OwnerReference{
+		APIVersion:         "v1",
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+		Kind:               "Node",
+		Name:               n.GetName(),
+		UID:                n.GetUID(),
+		Controller:         pointer.BoolPtr(true),
+	}))
+
+	if err := r.Create(context.TODO(), nm); err != nil {
+		klog.Error(err, " --> Problems in NamespaceMap creation")
+		return err
+	}
+
+	return nil
+}
+
+// first create a new NamespaceMap which preserves the Status and then delete the other
+func (r *VirtualNodeReconciler) regenerateNamespaceMap(nm mapsv1alpha1.NamespaceMap) error {
+	if slice.ContainsString(nm.GetFinalizers(), namespaceMapFinalizer, nil) {
+
+		// create a new namespaceMap with same NattingTable but with different Name
+		newNm := &mapsv1alpha1.NamespaceMap{}
+		nm.DeepCopyInto(newNm)
+		newNm.SetName("")
+		newNm.SetResourceVersion("")
+		newNm.SetGenerateName(fmt.Sprintf("%s-", nm.GetLabels()[const_ctrl.VirtualNodeClusterId]))
+
+		if err := r.Create(context.TODO(), newNm); err != nil {
+			klog.Error(err, " --> Problems in newNm NamespaceMap creation")
+			return err
+		}
+
+		if err := r.deleteNamespaceMap(nm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// This function manages NamespaceMaps Lifecycle on the basis of NamespaceMaps' number associated with the specific node
+func (r *VirtualNodeReconciler) namespaceMapLifecycle(n corev1.Node) error {
+
+	nms := &mapsv1alpha1.NamespaceMapList{}
+
+	if err := r.List(context.TODO(), nms, client.InNamespace(const_ctrl.MapNamespaceName), client.MatchingLabels{const_ctrl.VirtualNodeClusterId: n.GetAnnotations()[const_ctrl.VirtualNodeClusterId]}); err != nil {
+		klog.Error(err, " --> Unable to List NamespaceMaps")
+		return err
+	}
+
+	switch len(nms.Items) {
```

Comment:
What do you think about re-writing the logic in this switch statement? I think that in this case, it decreases the code readability. A further question: what if I have 3 `namespacemap` associated with the same node because of a bug, or a user error? You should enforce the uniqueness of a `namespacemap` in any case.

Reply:
yes, you are right, if cluster admin create more than one NamespaceMap beyond that created by virtual-node, this code wouldn't work. You suggest only to manage also this case, or to change the entire `switch `strategy, because in previous commit I have adopted an `if-else` strategy, but it wasn't clearer than this one...

Reply:
I would replace the switch statement with 3 `if` statements that check the length of the list (0, 1, >1). I think that the code readability would be improved.